### PR TITLE
Initial, partial 'Split' node implementation

### PR DESCRIPTION
**Pull request by mryndzionek**
I started working on this, but got sidetracked. Even if it's not yet complete, might be good to have it here for reference/history.

### CODE REVIEW

**Comment by mryndzionek** [comment]
The last commit is the initial Split's `print` implementation.
I tested "manually" using Torch's tensor split function.
After adding tests to `test/CMakeLists.txt`:

```
ONNX_backend_node_test(split_equal_parts_1d)
ONNX_backend_node_test(split_equal_parts_2d)
```

they fail with a segfault, because I assume the `split` input is present in `resolve`.
`split` input however is optional and not set in those tests.
So now this presents a problem - in order to implement the operation
I need to know the number of outputs. There are three possibilities:

 1. when `split` input is present the length of this tensor is the number of outputs
 2. in v18 on ONNX Split node has a `num_outputs` attribute
 3. the number of outputs is inferred from the whole model/graph

Option (3) seems to be the case for the two tests.
`get_number_of_outputs()` in `resolve` won't return this information, right?
The purpose of `resolve` is to register the outputs and only after that the number is known.
So how to solve this "chicken and egg problem"?

**Comment by mryndzionek** [comment]
Quick update. I hardcoded the number of outputs for the two tests and they pass. This gives me at least some confidence that my generated code works:
https://github.com/mryndzionek/onnx2c/commit/a29718e734199bf1aac420865814b60fcdbbca0b
https://github.com/mryndzionek/onnx2c/commit/b7faf11405ac776487970be6a726b907be12de81

**Comment by mryndzionek** [comment]
One more update. After adding tests with defined `split`:
```
ONNX_backend_node_test_singlefile(split_variable_parts_1d)
ONNX_backend_node_test_singlefile(split_variable_parts_2d)
ONNX_backend_node_test_singlefile(split_variable_parts_default_axis)
```
all three tests pass on this PR, so probably a good sign. Not sure what is the difference between `ONNX_backend_node_test` and `ONNX_backend_node_test_singlefile` (only the latter works).

**Comment by mryndzionek** [comment]
Would it be a good idea to pass `output_used.size()` to `resolve` [here](https://github.com/kraiskil/onnx2c/blob/687d5239abe0a49c273ccc7fb5408424cf6a0168/src/graph.cc#L338)?

**Comment by mryndzionek** [comment]
I rebased on master, refactored slightly and enabled passing tests. My previous question still stands - we need a way of inferring (at least) number of outputs from the graph and having this information available in `resolve`.

**Comment by kraiskil** [comment]
Wow, thank you for the effort!
I'll have to have a think about some of your questions - this project has been a bit on a hiatus for a while, and your questions go deep...

The question I can answer is the one about  `ONNX_backend_node_test_singlefile`. This CMake macro is the original one way of testing. It generates a single file from ONNX backend tests, where the onnx model and the test framework main file are generated into a single C file, and then compiled to a test. 

Some corner case test had a problem with this, and since it essentially is not what onnx2c does,  I wrote a new test macro `ONNX_backend_node_test` that generates two files. The model conversion is with onnx2c, and the test framework is generated with testgen into a separate file.

The latter is the "correct" way of doing the test, since it actually tests the onnx2c output without the interference of the test suite. 
But some tests remained that just wouldn't work with two separate files, so I left the old, singlefile, version in place in the CMake.

I'll try to get back to your questions soonest :)

**Comment by mryndzionek** [comment]
> But some tests remained that just wouldn't work with two separate files, so I left the old, singlefile, version in place in the CMake.

Yeah, so this is what I noticed. Some tests fail. For example slice: `./onnx2 onnx2c/onnx/onnx/backend/test/data/node/test_slice/model.onnx`. It fails because it accesses contents of tensor in `resolve`, but they are not initialized. `singlefile` seems to differ in this way that it initializes the tensors. Probably from the `.pb` files.

**Comment by kraiskil** [comment]
Right. This is the reason I decided to split up the test framework and generated model in two files.
When generated as a single file the "compile unit" includes the test cases buffers, i.e. the `split` tensors in this case.

But if you run the split file tests, you should see:
`onnx2c onnx/onnx/backend/test/data//node//test_split_variable_parts_1d/model.onnx` segfault. 
This is because the `Split::resolve()` now tries to access the data of the `split` tensor at "compile time" (i.e. `onnx2c` run time). But when the model is compiled in isolation, the `split` node's contents are not known. In this model, they are passed at inference time. 

Having the `split` as a model input is of course a bit silly example. But it demonstrates the case where `split` is not  a onnx-compile time constant, something the implementation should take into account.

**Comment by mryndzionek** [comment]
Thanks for the explanation. I also noticed that it's not uncommon for PyTorch to generate a graph like this:
Screenshot at 2024-08-09 14-09-56
This will now fail at `Expand` with:
```
[Fatal] (resolve) Unimplemented: Expand operand with non-constant input (shape).
onnx2c: onnx2c/src/nodes/expand.cc:53: virtual void toC::Expand::resolve(): Assertion `false' failed.
```
But from the graph we can see that technically the inputs are constant. There needs to be a transitive "constant propagation" mechanism implemented, I think.

**Comment by kraiskil** [comment]
That looks like a memory optimization, trading CPU power for ROM. Maybe pytorch has an option to not optimize for memory. 
Onnx2c isn't the cleanest tool to make optimizations on the graph level. Maybe onnx-sim or some similar tool would be a easier match to clean up the graph.

It could be done, but not only would the info about the input being constant needed, but also the result of the calculation, since the shape of the Expand output depends on this const value. I.e. each node Mul, Equal and Where should calculate their output at `resolve()` time. Which should not be impossible, since each of their input is const. Just a bit work and maybe a bit restructuring. (so definitely an issue/PR of its own :))

Slightly related, seems PyTorch 2.0 is going to be very trigger-happy wrapping layers in functions (see e.g. #46). With luck, that pattern of nodes would be wrapped in a function, which could maybe be easier to detect being compile time constant. I have started to look into how to implement onnx functions. I'll keep this use case in mind...

File: src/nodes/split.h [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	int64_t axis = 0;
+
+	virtual void parseAttributes( onnx::NodeProto &node ) override
+	{
+		for( const auto& a : node.attribute() ) {
+			if( a.name() == "axis" )
+				axis = parse_attribute_int(a);
+			else
```
Comment (kraiskil) [inline src/nodes/split.h:24]: Here I'd like the code to check for the other attributes too `num_outputs` and `split` (which was deprecated in ONNX Opset 13). It's ok to just error out with an "uninmplemented" if these are given - an explicit error is nicer than a silent failure.

File: src/nodes/split.h [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	{
+		auto split_sum = 0;
+
+		const Tensor *input = get_input_tensor(0);
+		const Tensor *split = get_input_tensor(1);
+		name_input(0, "input");
+		name_input(1, "split");
+
+		// TODO in v18 'num_outputs' is an attribute
+		int64_t num_outputs = split->data_dim[0];
```
Comment (kraiskil) [inline src/nodes/split.h:140]: `split` is optional, so it can be NULL here. Please guard against that.
As above, it's ok to error out with an "unimplemented".

Also, I'd add the check that if split exists, `split->isConst` is true, since othervise the graph has data-dependent tensor sizes (i.e. dynamic memory allocation would be needed).

**Review by kraiskil — COMMENTED** [review]
I think with the two requested changes, this is good to merge.

**Comment by mryndzionek** [comment]
Okay, I added the changes.

**Comment by kraiskil** [comment]
Thanks!
